README.md

# @chainsafe/netmask

Typescript implementation for using CIDR masks for address filtering.
Heavily inspired by go implementation.

## Features

* IPv4 and IPv6 support
* [Typescript](https://www.typescriptlang.org/) support

## How to use

`npm i -s @chainsafe/netmask`

or
`yarn add @chainsafe/netmask`

Example usage:

```typescript
import {IpNet} from "@chainsafe/netmask"

new IpNet("192.168.0.1/24").contains("192.168.0.16")
new IpNet("192.168.0.1", "24").contains("192.168.0.16")
new IpNet("2001:db8::/128").contains("2001:db8::")
new IpNet("192.168.0.1", "255.255.255.0").contains("192.168.0.16")
```

## Quick start

1. `yarn`
2. `yarn run build`
3. `yarn run test`


CHANGELOG.md
# Changelog

## [2.0.0](https://github.com/ChainSafe/netmask/compare/v1.0.0...v2.0.0) (2023-01-16)


### ⚠ BREAKING CHANGES

* rewritten api, using existing implementation of ip parsing

### Features

* rewritten api, using existing implementation of ip parsing ([cf5869c](https://github.com/ChainSafe/netmask/commit/cf5869c4fc6542f40833ac9b0f5dfe8629080ef5))

## 1.0.0 (2023-01-11)


### Features

* initial version ([439204c](https://github.com/ChainSafe/netmask/commit/439204c72dfc60727ab3dbd3620f82d6f7d8c1ea))


test/ip.spec.ts
import { parseIPv6 } from "@chainsafe/is-ip/parse";
import { expect } from "@esm-bundle/chai";
import { cidrMask } from "../src/cidr.js";
import {
  containsIp,
  IpNetRaw,
  iPv4FromIPv6,
  isIPv4mappedIPv6,
} from "../src/ip.js";

describe("ip", function () {
  describe("isIpv4MappedIpv6", function () {
    it("should be true", function () {
      expect(
        isIPv4mappedIPv6([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 127, 1, 2, 3])
      ).to.be.true;
    });
    it("should work", function () {
      expect(
        isIPv4mappedIPv6([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 254, 127, 1, 2, 3])
      ).to.be.false;
    });
  });

  describe("ipv4FromIpv6", function () {
    it("should be true", function () {
      expect(
        Array.from(
          iPv4FromIPv6(
            new Uint8Array([
              0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 127, 1, 2, 3,
            ])
          )
        )
      ).to.be.deep.equal([127, 1, 2, 3]);
    });
    it("should throw error", function () {
      expect(() =>
        Array.from(
          iPv4FromIPv6(
            new Uint8Array([
              0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 254, 127, 1, 2, 3,
            ])
          )
        )
      ).to.throw;
    });
  });

  describe("containsIP", function () {
    const cases: [Uint8Array | number[], IpNetRaw, boolean][] = [
      [
        [172, 16, 1, 1],
        { network: new Uint8Array([172, 16, 0, 0]), mask: cidrMask(12, 32) },
        true,
      ],
      [
        [172, 24, 0, 1],
        { network: new Uint8Array([172, 16, 0, 0]), mask: cidrMask(13, 32) },
        false,
      ],
      [
        [192, 168, 0, 3],
        {
          network: new Uint8Array([192, 168, 0, 0]),
          mask: new Uint8Array([0, 0, 255, 252]),
        },
        true,
      ],
      [
        [192, 168, 0, 4],
        {
          network: new Uint8Array([192, 168, 0, 0]),
          mask: new Uint8Array([0, 255, 0, 252]),
        },
        false,
      ],
      [
        parseIPv6("2001:db8:1:2::1")!,
        { network: parseIPv6("2001:db8:1::")!, mask: cidrMask(47, 128) },
        true,
      ],
      [
        parseIPv6("2001:db8:1:2::1")!,
        { network: parseIPv6("2001:db8:2::")!, mask: cidrMask(47, 128) },
        false,
      ],
      [
        parseIPv6("2001:db8:1:2::1")!,
        {
          network: parseIPv6("2001:db8:1::")!,
          mask: parseIPv6("ffff:0:ffff::")!,
        },
        true,
      ],
      [
        parseIPv6("2001:db8:1:2::1")!,
        {
          network: parseIPv6("2001:db8:1::")!,
          mask: parseIPv6("0:0:0:ffff::")!,
        },
        false,
      ],
    ];

    cases.forEach(([ip, ipNet, result]) => {
      it(`valid case ip=${String(ip)}, net=${String(
        ipNet.network
      )}, mask=${String(ipNet.mask)}`, function () {
        expect(containsIp(ipNet, ip)).to.be.equal(result);
      });
    });
  });
});


test/cidr.spec.ts
import { parseIPv4, parseIPv6 } from "@chainsafe/is-ip/parse";
import { expect } from "@esm-bundle/chai";
import { cidrMask, parseCidr } from "../src/cidr.js";
import { maskIp } from "../src/ip.js";

describe("cidr", function () {
  describe("parseCIDR", function () {
    const invalidCidr = [
      "192.168.1.1",
      "192.168.1.1/",
      "/8",
      // "192.168.1.1/255.255.255.0",
      // "192.168.1.1/35",
      // "2001:db8::1/-1",
      // "2001:db8::1/-0",
      // "-0.0.0.0/32",
      // "0.-1.0.0/32",
      // "0.0.-2.0/32",
      // "0.0.0.-3/32",
      // "0.0.0.0/-0",
      // "127.000.000.001/32",
    ];

    invalidCidr.forEach((cidr) => {
      it("invalid cidr - " + cidr, function () {
        expect(() => parseCidr(cidr)).to.throw("Failed to parse given CIDR");
      });
    });
  });

  describe("cidrMask", function () {
    const valid: [number, number, number[]][] = [
      [0, 32, [0, 0, 0, 0]],
      [12, 32, [255, 240, 0, 0]],
      [24, 32, [255, 255, 255, 0]],
      [32, 32, [255, 255, 255, 255]],
      [0, 128, [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]],
      [4, 128, [0xf0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]],
      [
        48,
        128,
        [0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
      ],
      [
        128,
        128,
        [
          0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
          0xff, 0xff, 0xff, 0xff, 0xff,
        ],
      ],
    ];

    valid.forEach(([ones, bits, result]) => {
      it(`valid case ones=${String(ones)}, bits=${String(bits)}`, function () {
        expect(Array.from(cidrMask(ones, bits))).to.be.deep.equal(result);
      });
    });

    const invalid = [
      [33, 32],
      [32, 33],
      [-1, 128],
      [128, -1],
    ];

    invalid.forEach(([ones, bits]) => {
      it(`invalid case ones=${String(ones)}, bits=${String(
        bits
      )}`, function () {
        expect(() => cidrMask(ones, bits)).to.throw();
      });
    });
  });

  describe("maskIP", function () {
    const testCases: [Uint8Array, Uint8Array, Uint8Array][] = [
      [
        new Uint8Array([192, 168, 1, 127]),
        new Uint8Array([255, 255, 255, 128]),
        new Uint8Array([192, 168, 1, 0]),
      ],
      [
        new Uint8Array([192, 168, 1, 127]),
        parseIPv4("255.255.255.192")!,
        new Uint8Array([192, 168, 1, 64]),
      ],
      [
        new Uint8Array([192, 168, 1, 127]),
        parseIPv6("ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffe0")!,
        new Uint8Array([192, 168, 1, 96]),
      ],
      [
        new Uint8Array([192, 168, 1, 127]),
        new Uint8Array([255, 0, 255, 0]),
        new Uint8Array([192, 0, 1, 0]),
      ],
      [
        parseIPv6("2001:db8::1")!,
        parseIPv6("ffff:ff80::")!,
        parseIPv6("2001:d80::")!,
      ],
      [
        parseIPv6("2001:db8::1")!,
        parseIPv6("f0f0:0f0f::")!,
        parseIPv6("2000:d08::")!,
      ],
    ];

    testCases.forEach(([ip, mask, result]) => {
      it(`valid case ip=${String(ip)}, mask=${String(mask)}`, function () {
        expect(maskIp(ip, mask)).to.be.deep.equal(result);
      });
    });

    const invalid = [
      [33, 32],
      [32, 33],
      [-1, 128],
      [128, -1],
    ];

    invalid.forEach(([ones, bits]) => {
      it(`invalid case ones=${String(ones)}, bits=${String(
        bits
      )}`, function () {
        expect(() => cidrMask(ones, bits)).to.throw();
      });
    });
  });
});


test/util.spec.ts
import { parseIP, parseIPv4 } from "@chainsafe/is-ip/parse";
import { expect } from "@esm-bundle/chai";
import { cidrMask } from "../src/cidr.js";
import { simpleMaskLength } from "../src/util.js";

describe("utils", function () {
  describe("simpleMaskLength", function () {
    const cases: [Uint8Array, number][] = [
      [cidrMask(26, 32), 26],
      [parseIPv4("255.0.255.0")!, -1],
      [parseIPv4("255.255.255.0")!, 24],
      [cidrMask(55, 128), 55],
      [parseIP("8000:f123:0:cafe::")!, -1],
    ];

    cases.forEach(([mask, length]) => {
      it(`mask=${String(mask)} result=${length}`, function () {
        expect(simpleMaskLength(mask)).to.be.equal(length);
      });
    });
  });
});


src/cidr.ts
import {
  IPv4,
  IPv4Len,
  IPv6,
  IPv6Len,
  maskIp,
  parseIPv4,
  parseIPv6,
} from "./ip.js";

export type Mask = number[];
export interface IPNet {
  net: IPv4 | IPv6;
  mask: Mask;
}

export function parseCidr(s: string): IPNet {
  const [address, maskString] = s.split("/");
  if (!address || !maskString)
    throw new Error("Failed to parse given CIDR: " + s);
  let ipLength = IPv4Len;
  let ip: IPv4 | IPv6 | null = parseIPv4(address);
  if (ip == null) {
    ipLength = IPv6Len;
    ip = parseIPv6(address);
    if (ip == null) throw new Error("Failed to parse given CIDR: " + s);
  }
  const m = parseInt(maskString, 10);
  if (
    Number.isNaN(m) ||
    String(m).length !== maskString.length ||
    m < 0 ||
    m > ipLength * 8
  ) {
    throw new Error("Failed to parse given CIDR: " + s);
  }
  const mask = cidrMask(m, 8 * ipLength);
  return {
    net: maskIp(ip, mask),
    mask,
  };
}

export function cidrMask(ones: number, bits: number): Mask {
  if (bits !== 8 * IPv4Len && bits !== 8 * IPv6Len)
    throw new Error("Invalid CIDR mask");
  if (ones < 0 || ones > bits) throw new Error("Invalid CIDR mask");
  const l = bits / 8;
  const m = new Array<number>(l).fill(0);
  for (let i = 0; i < l; i++) {
    if (ones >= 8) {
      m[i] = 0xff;
      ones -= 8;
      continue;
    }
    m[i] = 255 - (0xff >> ones);
    ones = 0;
  }
  return m;
}


src/index.ts
import { parseCidr } from "./cidr.js";
import { containsIp, parseIP } from "./ip.js";

export {
  parseIP,
  parseIPv4,
  parseIPv6,
  maskIp,
  iPv4FromIPv6,
  isIPv4mappedIPv6,
} from "./ip.js";
export type { IPv4, IPv6 } from "./ip.js";
export type { IPNet, Mask } from "./cidr.js";
export { parseCidr } from "./cidr.js";

/**
 * Checks if cidr block contains ip address
 * @param cidr ipv4 or ipv6 formatted cidr . Example 198.51.100.14/24 or 2001:db8::/48
 * @param ip ipv4 or ipv6 address Example 198.51.100.14 or 2001:db8::
 *
 */
export function cidrContains(cidr: string, ip: string): boolean {
  const netIP = parseCidr(cidr);
  const parsedIP = parseIP(ip);
  if (parsedIP == null) {
    throw new Error("Invalid ip");
  }
  return containsIp(netIP, parsedIP);
}
/**
 * Checks if network with given mask contains IP
 * @param cidr ipv4 or ipv6 formatted cidr . Example 198.51.100.14/24 or 2001:db8::/48
 * @param ip ipv4 or ipv6 address Example 198.51.100.14 or 2001:db8::
 *
 */
export function networkMaskContains(
  networkIp: string,
  mask: string,
  ip: string
): boolean {
  const parsedNetwork = parseIP(networkIp);
  const parsedMask = parseIP(mask);
  const parsedIP = parseIP(ip);
  if (parsedIP == null || parsedNetwork == null || parsedMask == null) {
    throw new Error("Invalid parameters");
  }
  return containsIp(
    {
      net: parsedNetwork,
      mask: parsedMask,
    },
    parsedIP
  );
}


src/ip.ts
import type { IPNet, Mask } from "./cidr.js";
import { allFF, deepEqual } from "./util.js";

export const IPv4Len = 4;
export const IPv6Len = 16;

export const maxIPv6Octet = parseInt("0xFFFF", 16);
export const ipv4Prefix = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255];

export type IPv4 = [number, number, number, number];
export type IPv6 = [
  number,
  number,
  number,
  number,
  number,
  number,
  number,
  number,
  number,
  number,
  number,
  number,
  number,
  number,
  number,
  number
];

export function maskIp(ip: IPv4 | IPv6, mask: Mask): IPv4 | IPv6 {
  if (mask.length === IPv6Len && ip.length === IPv4Len && allFF(mask, 0, 11)) {
    mask = mask.slice(12);
  }
  if (
    mask.length === IPv4Len &&
    ip.length === IPv6Len &&
    deepEqual(ip, ipv4Prefix, 0, 11)
  ) {
    ip = ip.slice(12) as IPv4;
  }
  const n = ip.length;
  if (n != mask.length) {
    throw new Error("Failed to mask ip");
  }
  const out = new Array<number>(n).fill(0) as IPv4 | IPv6;
  for (let i = 0; i < n; i++) {
    out[i] = ip[i] & mask[i];
  }
  return out;
}

export function containsIp(net: IPNet, ip: IPv4 | IPv6 | string): boolean {
  if (typeof ip === "string") {
    ip = parseIP(ip)!;
  }
  if (ip == null) throw new Error("Invalid ip");
  if (ip.length !== net.net.length) {
    return false;
  }
  for (let i = 0; i < ip.length; i++) {
    if ((net.net[i] & net.mask[i]) !== (ip[i] & net.mask[i])) {
      return false;
    }
  }
  return true;
}

export function parseIP(s: string): IPv4 | IPv6 | null {
  for (let i = 0; i < s.length; i++) {
    switch (s[i]) {
      case ".":
        return parseIPv4(s);
      case ":":
        return parseIPv6(s);
    }
  }
  return null;
}

export function isIPv4mappedIPv6(ip: IPv6): boolean {
  for (let i = 0; i < IPv6Len - IPv4Len; i++) {
    if (ip[i] !== ipv4Prefix[i]) return false;
  }
  return true;
}

export function iPv4FromIPv6(ip: IPv6): IPv4 {
  if (!isIPv4mappedIPv6(ip)) {
    throw new Error("Not a ipv4 mapped ipv6 address");
  }
  return [ip[12], ip[13], ip[14], ip[15]];
}

/**
 * Parses IPv4 address, returns ip address or null if invalid
 */
export function parseIPv4(s: string): IPv4 | null {
  const ip: IPv4 = [0, 0, 0, 0];
  if (s.length === 0) return null;
  const parts = s.split(".");
  if (parts.length < IPv4Len) return null;
  for (let i = 0; i < IPv4Len; i++) {
    const part = parts[i];
    if (part.startsWith("-")) return null;
    const p = parseInt(parts[i], 10);
    if (Number.isNaN(p)) return null;
    if (p === 0 && part.length > 1) return null;
    if (p < 0 || p > 255) return null;
    if (p > 1 && parts[i].startsWith("0")) {
      return null;
    }
    ip[i] = p;
  }
  return ip;
}

/**
 * parseIPv6 parses s as a literal IPv6 address described in RFC 4291
 * and RFC 5952. Returns either IPV6 address or null if invalid
 */
export function parseIPv6(s: string): IPv6 | null {
  const ipv6: IPv6 = new Array<number>(16).fill(0) as IPv6;
  let elipsis = -1;
  if (s.length > 2 && s[0] === ":" && s[1] === ":") {
    elipsis = 0;
    s = s.slice(2);
    if (s.length === 0) {
      return ipv6;
    }
  }
  let i = 0;
  while (i < IPv6Len) {
    let colonIndex = s.indexOf(":");
    let dotIndex = s.indexOf(".");
    colonIndex = colonIndex == -1 ? Infinity : colonIndex;
    dotIndex = dotIndex == -1 ? Infinity : dotIndex;
    const part = s.slice(0, Math.min(colonIndex, dotIndex));
    if (!part.match(/^[A-Fa-f0-9]*$/)) return null;
    //parse hex
    const p = parseInt(part, 16);
    if (Number.isNaN(p) || p < 0 || p > maxIPv6Octet) return null;
    // If followed by dot, might be in trailing IPv4.
    if (s[part.length] === ".") {
      if (elipsis < 0 && i != IPv6Len - IPv4Len) {
        //not the right place
        return null;
      }
      if (i + IPv4Len > IPv6Len) {
        // Not enough room.
        return null;
      }
      const ip4 = parseIPv4(s);
      if (ip4 == null) return null;
      ipv6[i] = ip4[0];
      ipv6[i + 1] = ip4[1];
      ipv6[i + 2] = ip4[2];
      ipv6[i + 3] = ip4[3];
      s = "";
      i += IPv4Len;
      break;
    }
    ipv6[i] = p >> 8;
    ipv6[i + 1] = p & 0xff;
    i += 2;
    s = s.slice(part.length);
    if (s.length === 0) break;
    if (s[0] !== ":" || s.length === 1) {
      return null;
    }
    s = s.slice(1);
    if (s[0] === ":") {
      if (elipsis >= 0) return null;
      elipsis = i;
      s = s.slice(1);
      if (s.length === 0) break;
    }
  }
  if (s.length !== 0) {
    return null;
  }
  // If didn't parse enough, expand ellipsis.
  if (i < IPv6Len) {
    if (elipsis < 0) {
      return null;
    }
    const n = IPv6Len - i;
    for (let j = i - 1; j >= elipsis; j--) {
      ipv6[j + n] = ipv6[j];
    }
    for (let j = elipsis + n - 1; j >= elipsis; j--) {
      ipv6[j] = 0;
    }
  } else if (elipsis >= 0) {
    // Ellipsis must represent at least one 0 group.
    return null;
  }
  return ipv6;
}
